README.md
<img src="https://raw.githubusercontent.com/anxdpanic/plugin.video.composite_for_plex/master/icon.png" width="256" height="256" alt="Composite">

# Composite

[![Build Status](https://img.shields.io/endpoint.svg?url=https%3A%2F%2Factions-badge.atrox.dev%2Fanxdpanic%2Fplugin.video.composite_for_plex%2Fbadge&logo=none)](https://actions-badge.atrox.dev/anxdpanic/plugin.video.composite_for_plex/goto)
![License](https://img.shields.io/badge/license-GPL--2.0--or--later-success.svg)
![Kodi Version](https://img.shields.io/badge/kodi-jarvis%2B-success.svg)
![Contributors](https://img.shields.io/github/contributors/anxdpanic/plugin.video.composite_for_plex.svg)

Composite allows media and metadata stored in the Plex Media Server (PMS) to be viewed and played using the Kodi 16+ interface.

- Installation
    -
    ##### Kodi 18+
    * [Install from official repository](https://kodi.wiki/view/HOW-TO:Install_add-ons)
    
    ---
    
    ##### Kodi 16+
    * Kodi 17+: Enable - `Settings -> System -> Add-ons -> Unknown Sources`
    1. Download repository 
        - Kodi 16-18: [repository.anxdpanic-x.x.x.zip](https://panicked.xyz/repositories/repository.anxdpanic-2.0.0.zip)
        - Kodi 19: [repository.anxdpanic-x.x.x+matrix.1.zip](https://panicked.xyz/repositories/matrix/repository.anxdpanic-2.0.0+matrix.1.zip)
    2. [Install from zip file](http://kodi.wiki/view/Add-on_manager#How_to_install_from_a_ZIP_file) (repository.anxdpanic-x.x.x.zip)
    3. [Install from repository](http://kodi.wiki/view/add-on_manager#How_to_install_add-ons_from_a_repository) (anxdpanic Add-on Repository)

- Usage
    -

    Composite should work "out of the box" in most cases, as the default allows for automatic server discovery.
    If this doesn't work, then discovery can be switched off, and a manually entered hostname or IP address can be used.

    If you are watching remotely, or on a bandwidth limited network, switch on Transcoding to reduce the media quality
    to one that works best for you.

- Kodi 18+ Library
    -
    
    To add Composite (Plex) media to your library, add a video source for movies/tv shows.
 
    - Will not have support for most Plex features.
    
    #### Sources
   
    - **Movies:** `plugin://plugin.video.composite_for_plex/library/movies/`
      - Choose information provider: Local information only
      - Movies are in separate folders that match the movie title: Disabled
      - Scan recursively: Disabled
    - **TV Shows:** `plugin://plugin.video.composite_for_plex/library/tvshows/`
      - Choose information provider: Local information only
      - Selected folder contains a single TV show: Disabled

    **Wiki:** https://kodi.wiki/view/Adding_video_sources <br/>
    Instead of Steps 4-5 add the appropriate `plugin://` url from the sources above <br/>
    Instead of Step 9 use Local information only

- TraktToKodi
    -
    
    This add-on can be paired with the TraktToKodi web browser extension. <br/>
    
    In your TraktToKodi profile, set the `Add-on ID` to `plugin.video.composite_for_plex`

    **Chrome:**
    - Install from [Chrome Web Store](https://chrome.google.com/webstore/detail/trakttokodi/jongfgkokmlpdekeljpegeldjofbageo) <br/>
    - https://github.com/anxdpanic/TraktToKodi-Extension/tree/chrome
    
    **Firefox:**
    - Install from [AMO Gallery](https://addons.mozilla.org/en-US/firefox/addon/trakttokodi/)
    - https://github.com/anxdpanic/TraktToKodi-Extension/tree/firefox

- Support
    -
  
    Post an [issue](https://github.com/anxdpanic/plugin.video.composite_for_plex/issues)  
    Post on the [forum](https://forum.kodi.tv/showthread.php?tid=351179) 
---

Translation
-
    
Thanks to **Maven85** (de_DE)

Artwork
-
    
Thanks to **spooknik** (Icon and Fanart)
    
--- 
This add-on is a fork of [PleXBMC by Hippojay](https://github.com/hippojay/plugin.video.plexbmc)



changelog.txt
1.4.2
[fix] use of deprecated ElementTree.getiterator
[fixup] custom access urls
[add] Detect Servers to main menu
[chg] only refresh server cache on connection errors
[upd] linting
[lang] update translations from Weblate

1.4.1
[fix] xbmc.translatePath -> xbmcvfs.translatePath |Kodi 19|
[upd] improve resuming playback
[upd] improve accuracy of library context menus |Kodi 19|
[chg] group playlist sections on main menu
[chg] always used cached servers during playback

1.4.0
[add] Combined Sections to main menu
[add] Composite Playlist to main menu, create a playlist based the information you provide
[add] Composite context menu for library and widgets
    - adds Play Transcoded, Mark as un/watched
[add] Section selection for Kodi Library support
[add] settings to included watched content in combined 'Recently Added Episodes/Movies' sections
[chg] Remove HTTPS from settings and move Certificate Verification from settings to Manage Servers

1.3.0
[add] menu items 'TV Shows/Movies on Deck' and widgets 'All Servers: TV Shows/Movies on Deck'
    for on deck results from all servers
[add] menu items 'Recently Added Episodes/Movies' and widgets
    'All Servers: Recently Added Episodes/Movies' for recently added results from all servers
[chg] movie collections default to sorting by year
[fix] 'Always Transcode' setting is now applied
[fix] minor fixups
[upd] use new settings format for Kodi 19

1.2.0
[add] added intro skipping (Requires Plex Pass)
[add] added lyrics info label for lyrics provided by Plex (Requires Plex Pass)
[add] added setting to configure forced subtitles as default(forced)
[add] added setting to configure mixed content type behavior
[add] added support for custom access urls (ie. nginx)
          - Main Menu - Manage Servers - <server> - Custom access urls - Add custom access url
[fix] possible timestamp out of range for movies
[chg] All Movies, All Shows, All Artists, and All Photos are no longer shortened to All
[lang] German (de_DE) |contrib: Maven85|

1.1.1
[fixup] logging

1.1.0
[add] end-point to search for movies/episodes from all available servers for external use ie. TheMovieDb Helper
[fix] browsing By Folder
[fix] login on Kodi 19/Matrix due to removal of deprecated isPassword
[fix] Up Next when duration is unavailable
[upd] improve debug logging and masking of private data

1.0.0
[add] new fanart and icon |contrib: spooknik|
[upd] include more artwork when creating the playable item

0.8.3
[fix] resuming playback
[add] Settings - Look and Feel - Episode sort method: Kodi/Plex

0.8.2
[fix] listing of playlists in some cases |contrib: Maven85|
[lang] German (de_DE) |contrib: Maven85|

0.8.1
[upd] updates for official repository review

0.8.0
[fix] incorrect next episode sent to Up Next when playing merged episodes ie. S01E01-E02
[upd] Up Next now handles starting playback of the next episode
[upd] improve plex identification |contrib: Maven85|
[upd] refactoring
[upd] linting / cosmetics

0.7.1
[add] add cast to tv shows
[fix] set content type for library endpoints
[upd] linting / cosmetics

0.7.0
[add] Setting to configure service.upnext notification encoding (hex/base64)
[upd] Simplify hash generation for library integration of tv shows
[upd] Linting
[upd] Refactoring

0.6.0
[upd] cosmetics

0.5.0
[add] Playlist creation and deletion
[add] Support photo playlists
[fix] add basic path protection to cache related actions
[fix] only allow one dialog at a time ie. accidentally opening multiple sign in dialogs
[fix] Settings - Server - Manage myPlex when sign in required
[fix] Photos not playing
[upd] improve Manage myPlex and Sign In dialogs
[upd] simplify the main menu when not signed in
[rem/fix] remove redirect hack that was used for signing in
[lang] German (de_DE) |contrib: Maven85|

0.4.0
[add] Plex Companion receiver *experimental
[add] Settings - Look and Feel - Main Menu - Show [menu item]
    - show or hide some main menu items
[fix] Settings not taking effect due to reuselanguageinvoker
[fix] Only verify cert for secure connections, and honor the Certificate Verification setting |contrib: Maven85|
[fix] Some sockets not being closed after use
[upd] Updated the settings menu
[rem] Remove Clean context menus setting
[lang] German (de_DE) |contrib: Maven85|

0.3.0
[add] Kodi 18+ library integration support
    - Movies: add video source - plugin://plugin.video.composite_for_plex/library/movies/
    - TV Shows: add video source - plugin://plugin.video.composite_for_plex/library/tvshows/
    *** https://kodi.wiki/view/Adding_video_sources
    *** Will not have support for most Plex features
[add] Settings - Cache - Data Cache, to enable or disable the use of data caching
[upd] Improve device identification |contrib: Maven85|
[upd] Additional list item information at playback |contrib: Maven85|
[upd] Improve the speed of cache deletion
[upd] refactor items and processing into smaller pieces
[upd] refactor common, constants, and logging
[lang] German (de_DE) |contrib: Maven85|

0.2.1
[fix] TraktToKodi not starting playback of some episodes
[upd] TraktToKodi title comparisons

0.2.0
[add] support for TraktToKodi browser extension
    - https://chrome.google.com/webstore/detail/trakttokodi/jongfgkokmlpdekeljpegeldjofbageo
    - https://addons.mozilla.org/en-US/firefox/addon/trakttokodi/

0.1.4
[add] translations for widget titles
[fix] data cache not updated on mark (un)watched, and playback
[upd] improve data caching, less cached data required
[upd] refactor data caching
[upd] refactor list item creation

0.1.3
[chg] use same caching mechanism for data caching as server cache
    - [fix] widget caching
[rem] dependency on script.common.plugin.cache

0.1.2
[fixup] wrong minimum dependency version

0.1.1
[add] Widgets
[add] Data caching using script.common.plugin.cache
[fix] Episode artwork

0.1.0
[upd] refactor routing and add-on
[upd] cosmetics

0.0.10
[upd] update Sign In and Manage myPlex dialog layouts
[fix] playback monitoring threads not being cleaned up

0.0.9

resources/lib/common.py
import inspect
import os
import sys
import socket
import re

import xbmc
import xbmcaddon

from .settings import AddonSettings


class PrintDebug:

    def __init__(self, main, sub=None):

        self.main = main
        if sub:
            self.sub = "." + sub
        else:
            self.sub = ''

        self.level = settings.get_debug()
        self.privacy = settings.get_setting('privacy')

        self.DEBUG_OFF = 0
        self.DEBUG_INFO = 1
        self.DEBUG_DEBUG = 2
        self.DEBUG_DEBUGPLUS = 3
        self.DEBUG_HELPER = 4
        self.LOG_ERROR = 9
        self.token_regex = re.compile('-Token=[a-z|0-9].*?[&|$]')
        self.ip_regex = re.compile('\.\d{1,3}\.\d{1,3}\.')
        self.user_regex = re.compile('-User=[a-z|0-9].*?[&|$]')

        self.DEBUG_MAP = {self.DEBUG_OFF: "off",
                          self.DEBUG_INFO: "info",
                          self.DEBUG_DEBUG: "debug",
                          self.DEBUG_DEBUGPLUS: "debug+",
                          self.DEBUG_HELPER: "debug+h",
                          self.LOG_ERROR: 'error'}

    def get_name(self, level):
        return self.DEBUG_MAP[level]

    def error(self, message):
        return self.__print_message(message, 9)

    def warn(self, message):
        return self.__print_message(message, 0)

    def info(self, message):
        return self.__print_message(message, 1)

    def debug(self, message):
        return self.__print_message(message, 2)

    def dev(self, message):
        return self.__print_message(message, 3)

    def debugplus(self, message):
        return self.__print_message(message, 3)

    def debug_helper(self, message):
        return self.__print_message(message, 4)

    def __print_message(self, msg, level=1):
        if self.privacy:
            msg = self.token_regex.sub("X-Plex-Token=XXXXXXXXXX&", str(msg))
            msg = self.ip_regex.sub(".X.X.", msg)
            msg = self.user_regex.sub("X-Plex-User=XXXXXXX&", msg)

        if self.level >= level != self.LOG_ERROR:
            try:
                xbmc.log("%s%s -> %s : %s" % (self.main, self.sub, inspect.stack(0)[2][3], msg.encode('utf-8')), xbmc.LOGDEBUG)
            except:
                xbmc.log("%s%s -> %s : %s [NONUTF8]" % (self.main, self.sub, inspect.stack(0)[2][3], msg), xbmc.LOGDEBUG)
        elif level == self.LOG_ERROR:
            try:
                xbmc.log("%s%s -> %s : %s" % (self.main, self.sub, inspect.stack(0)[2][3], msg.encode('utf-8')), xbmc.LOGERROR)
            except:
                xbmc.log("%s%s -> %s : %s [NONUTF8]" % (self.main, self.sub, inspect.stack(0)[2][3], msg), xbmc.LOGERROR)

    def __call__(self, msg, level=1):
        return self.__print_message(msg, level)


def get_platform():
    if xbmc.getCondVisibility('system.platform.osx'):
        return "OSX"
    elif xbmc.getCondVisibility('system.platform.atv2'):
        return "ATV2"
    elif xbmc.getCondVisibility('system.platform.ios'):
        return "iOS"
    elif xbmc.getCondVisibility('system.platform.windows'):
        return "Windows"
    elif xbmc.getCondVisibility('system.platform.linux'):
        return "Linux/RPi"
    elif xbmc.getCondVisibility('system.platform.android'):
        return "Linux/Android"
    return "Unknown"


def wake_servers():
    if settings.get_setting('wolon'):
        from WOL import wake_on_lan
        log_print = PrintDebug("PleXBMC", "wake_servers")
        log_print.debug("PleXBMC -> Wake On LAN: true")
        for servers in settings.get_wakeservers():
            if servers:
                try:
                    log_print.debug("PleXBMC -> Waking server with MAC: %s" % servers)
                    wake_on_lan(servers)
                except ValueError:
                    log_print.debug("PleXBMC -> Incorrect MAC address format for server %s" % servers)
                except:
                    log_print.debug("PleXBMC -> Unknown wake on lan error")


def setup_python_locations():
    setup = {'__addon__': xbmcaddon.Addon()}
    setup['__cachedir__'] = setup['__addon__'].getAddonInfo('profile')
    setup['__cwd__'] = xbmc.translatePath(setup['__addon__'].getAddonInfo('path')).decode('utf-8')
    setup['__version__'] = setup['__addon__'].getAddonInfo('version')
    setup['__resources__'] = xbmc.translatePath(os.path.join(setup['__cwd__'], 'resources', 'lib'))
    sys.path.append(setup['__resources__'])
    return setup


def is_ip(address):
    """from http://www.seanelavelle.com/2012/04/16/checking-for-a-valid-ip-in-python/"""
    try:
        socket.inet_aton(address)
        ip = True
    except socket.error:
        ip = False

    return ip


def get_platform_ip():
    return xbmc.getIPAddress()


GLOBAL_SETUP = setup_python_locations()
GLOBAL_SETUP['platform'] = get_platform()
GENERIC_THUMBNAIL = xbmc.translatePath('special://home/addons/plugin.video.plexbmc/resources/media/thumb.png')
REQUIRED_REVISION = "1.0.7"
settings = AddonSettings('plugin.video.plexbmc')

# Get the setting from the appropriate file.
MODE_GETCONTENT = 0
MODE_TVSHOWS = 1
MODE_MOVIES = 2
MODE_ARTISTS = 3
MODE_TVSEASONS = 4
MODE_PLAYLIBRARY = 5
MODE_TVEPISODES = 6
MODE_PLEXPLUGINS = 7
MODE_PROCESSXML = 8
MODE_CHANNELSEARCH = 9
MODE_CHANNELPREFS = 10
MODE_PLAYSHELF = 11
MODE_BASICPLAY = 12
MODE_SHARED_MOVIES = 13
MODE_ALBUMS = 14
MODE_TRACKS = 15
MODE_PHOTOS = 16
MODE_MUSIC = 17
MODE_VIDEOPLUGINPLAY = 18
MODE_PLEXONLINE = 19
MODE_CHANNELINSTALL = 20
MODE_CHANNELVIEW = 21
MODE_PLAYLIBRARY_TRANSCODE = 23
MODE_DISPLAYSERVERS = 22
MODE_MYPLEXQUEUE = 24
MODE_SHARED_SHOWS = 25
MODE_SHARED_MUSIC = 26
MODE_SHARED_PHOTOS = 27
MODE_DELETE_REFRESH = 28
MODE_SHARED_ALL = 29
MODE_PLAYLISTS = 30

SUB_AUDIO_XBMC_CONTROL = "0"
SUB_AUDIO_PLEX_CONTROL = "1"
SUB_AUDIO_NEVER_SHOW = "2"
